### ISSUE TITLE
Review the MFS "cache" (it's not a cache)

### ISSUE THREAD

**Issue by schomatis** [ipfs/go-mfs#54]
The structure inside `Directory` named [`entriesCache`](https://github.com/ipfs/go-mfs/blob/6c0036189ea0941b3f88ad62ba76fb1748e2a266/dir.go#L30-L32) keeps a record of the `Directory`s and `File`s structures created when accessing an entry in MFS, we've been calling this a "cache" but its semantics far exceeds that functionality and an incorrect use of it can lead to severe inconsistencies in the filesystem.

## Background

We don't have a single point of entry to interact with MFS. The `Root` (a glorified `Directory` with some special operations associated to it), although the main controller, it's not the only one. The user of the API can request internal `Directory`s and `File`s, that is, not only a `FileDescriptor` that tells MFS which file it wants to modify but rather the internal MFS representation of that file (i.e., the `File` structure).

(Beyond the scope of the current issue, note for future references: I think we should redesign the API to disallow that, making `Directory`s and `File`s private and only issuing references to them, like file descriptors or string paths.)

That `File` the user can hold has a reference to its parent (necessary because of the way a content-addressed system works) that allows it to modify other directories in the hierarchy (through `updateChildEntry` of the [`parent`](https://github.com/ipfs/go-mfs/blob/6c0036189ea0941b3f88ad62ba76fb1748e2a266/root.go#L47) interface).

Even if a `File` is removed from the MFS (i.e., [`Unlink`ed](https://github.com/ipfs/go-mfs/blob/6c0036189ea0941b3f88ad62ba76fb1748e2a266/dir.go#L330)) that `File` is still "connected" to its parent (and hence to the rest of the MFS), and a modification to that `File` (and a posterior `Flush`) would end up modifying the MFS root (and hence changing the file hierarchy state), for a simple test example see https://github.com/ipfs/go-mfs/pull/52/commits/cb7c5db8b5daf5c115e695db6ff3056cb153e582. (Technically the `Unlink` removes the file from the UnixFS layer but the `parent` keeps the connection at the MFS layer, overrunning the former when the MFS is synced).

## The directory "cache"

Beyond the simple cache dynamics of keeping `FSNode`s retrieved from the UnixFS layer available in the `Directory` structure, the `entriesCache` structure can also be thought of as a control structure that keeps records of issued `Directory`s and `File`s structures to the user to avoid duplicating the `FSNode` references to a specific path component in the MFS.

Why do we need to avoid duplication? For the same problem described above, e.g., we only allow one write-mode file descriptor per file at a time so there are not two simultaneous sources of modification for the same file that would cause inconsistencies (basically there would be two completely different versions and whichever `File` is `Flush`ed last "wins", overwriting the other in the entry of the parent's directory); the inconsistency arises because we've issued structures that have references (`parent`) to internal components of our MFS representation. (A very similar case can be made for `Directory`, with the added complexity that we don't have "directory descriptors" that enforce a single source of modification).

To make matters more complicated, an [`Uncache`](https://github.com/ipfs/go-mfs/blob/6c0036189ea0941b3f88ad62ba76fb1748e2a266/dir.go#L205) method was added to the API to allow the user to just remove an entry from `entriesCache` without any type of control (the motivation was to avoid high memory usage when listing, i.e., opening every entry of, an entire MFS, see https://github.com/ipfs/go-ipfs/pull/3027#issuecomment-448678709). This would open another inconsistency path similar to the case described above but calling `Uncache` instead of `Unlink`.

## Proposed solution

(By no means the optimal one.) Enforce a disconnection at the MFS level (since this layer "dominates" UnixFS, it's the one that counts in this context). The simple way seems to be through the `parent` reference (e.g., setting it to `nil`). That is, when `Unlink`ing, `Uncache`ing and similar disconnect the `FSNode` from that `parent` so any future modifications won't be reflected in the MFS. It's important to note that that file/directory is still valid, i.e., it's still a coherent UnixFS file or directory (with it's associated children), just that it's not *part* of the MFS in the particular directory entry it was registered, but that `File` or `Directory` can just as well be moved to a different MFS or to a different path in the same MFS it was disconnected from, that is, we could still operate on it as a separate entity, e.g., a "disconnected" directory could even be transformed to a new MFS feeding its UnixFS node to `NewRoot`.

See this [patch](https://github.com/ipfs/go-mfs/pull/52/files/cb7c5db8b5daf5c115e695db6ff3056cb153e582..7d55d64c07751e9bc6bd4dfe7b2077eeb938649b) as an incomplete and incorrect PoC that attempted to move in that direction.

Probably the discussed "disconnection" logic should be part of `Uncache` and everyone else (e.g., `Unlink`) should call it when removing/renaming/etc an entry.

The downside of this approach is that we'll be modifying the `ìnode` structure (containing `parent`) *after* creation, something that it was not designed for. Nevertheless this is a simple enough structure so it can be redesigned with that in mind (and probably the `parent` should be extracted outside since it carries semantics quite different from the rest of the `ìnode` variables). Another alternative could be to add an explicit `unlinked` flag, to avoid modifying other parts of the `FSNode` structure, that would not permit any `Flush` operation to be propagated upward. Either way this "disconnection" operation will need to be lock-protected.

The `entriesCache` definitely needs to be renamed and we should drop the "cache" terminology altogether throughout the code since it gives it a benign appearance to a control structure that if not handled properly can cause serious inconsistencies.

More tests need to be created to explore this inconsistency vector, what we need to check is that there aren't two `FSNode`s that can cause simultaneous modifications to the same directory, e.g., there shouldn't be two "active" (i.e., not "disconnected") `File`s at the same moment for the same path, a renamed file/directory shouldn't also suffer from this problem.

**Comment by schomatis** [comment]
Assigned to @nitishm.

Labeling as `blocked` pending a review and feedback from @Stebalien on this issue (that is, @nitishm wait for @Stebalien to check this logic before starting to code).

**Comment by Stebalien** [comment]
Thanks for writing this up. That background should be helpful to anyone trying to tackle this issue.

WRT the proposed solution, it helps by actually exposing the error to the user instead of failing silently but `Uncache` can never be safe with multiple callers. One caller can still clobber another by calling `Uncache` before the other flushes. Really, we need reference counting and a `Close` method on directories/files.

> (Beyond the scope of the current issue, note for future references: I think we should redesign the API to disallow that, making Directorys and Files private and only issuing references to them, like file descriptors or string paths.)

So, I agree we need a clear separation of private datastructures versus public datastructures. However, these types *are* effectively handles, at the moment. The key issue is that we don't provide any way to *close* them.

**Comment by nitishm** [comment]
Hey hey. @schomatis @stebalien Any updates on this issue ?

**Comment by schomatis** [comment]
Hey @nitishm! Sorry for the delay, I'm unblocking this issue, please coordinate with @Stebalien on how to move forward. If you feel comfortable with the information provided here you can draft a PR to start addressing this issue, I still think there should be some more designing discussion going on here.

**Comment by Stebalien** [comment]
Personally, I recommend taking a look at how we use MFS (take a look at the adder, the `ipfs files` commands, and at https://github.com/ipfs/go-ipfs/issues/5003) and then re-desigining the interfaces from scratch. From what I've seen, MFS isn't something we can fix with a few band aids. 

Non-exhaustive list of issues:

* Does not play well with moving/copying files with concurrent writers. That is, directories are never "locked" for writing.
* Does not have atomic rename.
* Leaks memory over time due to the not-cache (this issue).
* Doesn't allow concurrent writers to the same file. That is, every writer effectively gets a full file lock.
* Constantly recomputes hashes, even when we haven't requested them yet. That is, there's no way to "flush" changes to make these changes visible to other readers without also recomputing a bunch of hashes.


However, I understand that this is probably a _lot_ of work. I'm just concerned that continuing to work on MFS as-is will have diminishing returns. @djdv any thoughts on how this relates to FUSE?

**Comment by djdv** [comment]
>any thoughts on how this relates to FUSE?

In general I think there's a lot of overlap in these issues, however I don't know what (if anything) from [`mount`](https://github.com/djdv/go-ipfs/tree/feat/win-mount) is applicable here.
Specifically, I'm currently trying to implement async IPNS write operations and encountering these same problems. This requires some thought around something like a key-root continuity manager.
In essence it feels like I'm reimplimenting an async form of MFS since I'm doing the same mappings (paths to some dynamic root hash; it's even worse since IPNS roots can be files). However it remains to be seen how this will be done, and more importantly, if it can be done in an optimal manner.

**Comment by aschmahmann** [comment]
This repository is no longer maintained and has been copied over to [Boxo](https://github.com/ipfs/boxo).  In an effort to avoid noise and crippling in the Boxo repo from the weight of issues of the past, we are closing most issues and PRs in this repo.  Please feel free to open a new issue in Boxo (and reference this issue) if resolving this issue is still critical for unblocking or improving your usecase.

You can learn more in the [FAQs for the Boxo repo copying/consolidation effort](https://github.com/ipfs/boxo/wiki/Copied-or-Migrated-Repos-FAQ).
